### PR TITLE
Feat: Adding Pushwoosh Provider

### PR DESCRIPTION
**Pull request by CursedRock17**
- **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Feature: Added the Pushwoosh Push Provider
- **Why was this change needed?** (You can also link to an open issue here)
It was just another Provider that needed to be added although there isn't an official issue for it to be added. This is a good place to start for adding this towards the repository as all of the basic information is there.

- **Other information**:
Currently missing some of the additional extra features but those can come in the future. Although it's almost finished I didn't add the Logos and Documentation because I don't know if we needed rights to the logos, but that's an easy fix

### CODE REVIEW

**Comment by p-fernandez** [comment]
Would you mind to open the issue also? It is a great contribution and adding the issue of request to add the provider will help us to keep track of it. Next time before working on it please open the issue first and we will assign it to you so you can work on it without potentially clashing with anyone else interested in working on it.

File: README.md [hunk 1]
```diff
@@ -14,12 +14,10 @@
 </br>
 
 <div align="center">
-  <a href="https://novu.co" target="_blank">
```
Comment (p-fernandez) [inline README.md:17]: The changes in this file shouldn't belong to this PR. Looks like you branched off from an old version of the codebase.

File: README.md [hunk 2]
```diff
@@ -158,14 +152,14 @@ Novu provides a single API to manage providers across multiple channels with a s
...
 - [ ] RingCentral
 
-#### 📱 Push 
+#### 📱 Push
 
 - [x] [FCM](https://github.com/novuhq/novu/tree/main/providers/fcm)
 - [x] [Expo](https://github.com/novuhq/novu/tree/main/providers/expo)
 - [ ] [SNS](https://github.com/novuhq/novu/tree/main/providers/sns)
-- [ ] Pushwoosh
+- [ ] [Pushwoosh](https://github.com/novuhq/novu/tree/main/providers/pushwoosh)
```
Comment (p-fernandez) [inline README.md:160]: Before adding this change you need to update the branch to the current latest changes of our codebase.

File: apps/api/src/app/events/services/push-service/handlers/pushwoosh.handler.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+      triggeredFunction: credentials.triggeredFunction,
+    });
+  }
+}
+
+interface IPushwooshConfig {
+  applicationCode?: string;
+  defaultNotificationTitle?: string;
+  autoSubscribe?: boolean;
+  triggeredFunction?: any;
```
Comment (p-fernandez) [inline apps/api/src/app/events/services/push-service/handlers/pushwoosh.handler.ts:47]: We can type functions if this property is expected to be a function.

File: libs/dal/src/repositories/integration/integration.entity.ts [hunk 4]
```diff
@@ -19,6 +19,13 @@ export interface ICredentials {
   clientId?: string;
   projectName?: string;
   serviceAccount?: string;
+  applicationCode?: string;
+  defaultNotificationTitle?: string;
+  triggeredFunction?: any;
+  autoSubscribe?: boolean;
+  userId?: string;
+  tags?: object;
+  defaultNotificationImageURL?: string;
```
Comment (p-fernandez) [inline libs/dal/src/repositories/integration/integration.entity.ts:28]: @scopsy @davidsoderberg @ainouzgali @djabarovgeorge this entity is growing a lot. I would add it to technical debt to find a maintainable solution.

File: libs/shared/src/consts/providers/provider.enum.ts [hunk 5]
```diff
@@ -55,6 +61,7 @@ export enum PushProviderIdEnum {
   FCM = 'fcm',
   APNS = 'apns',
   EXPO = 'expo',
+  Pushwoosh = 'pushwoosh',
```
Comment (p-fernandez) [inline libs/shared/src/consts/providers/provider.enum.ts:64]:
```suggestion
  PUSHWOOSH = 'pushwoosh',
```
Please keep the consistency by following the Enum standard.

File: providers/pushwoosh/README.md [hunk 6]
```diff
@@ -0,0 +1,23 @@
+## Usage
+
+-- First, you must create the files that are necessary to exmply push notifications. When Creating the files, manifest.json and pushwoosh-service-worker.js must be at the top level root. Then the returned string and paste in the head elemnt of the index.html file
```
Comment (p-fernandez) [inline providers/pushwoosh/README.md:3]: Seems there are some typos here.

File: providers/pushwoosh/src/lib/pushwoosh.provider.spec.ts [hunk 7]
```diff
@@ -0,0 +1,3 @@
+import { PushwooshPushProvider } from './pushwoosh.provider';
+
+test('should trigger pushwoosh library correctly', async () => {});
```
Comment (p-fernandez) [inline providers/pushwoosh/src/lib/pushwoosh.provider.spec.ts:3]: I guess this is work to do as some testing would be expected.

File: providers/pushwoosh/src/lib/pushwoosh.provider.ts [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+      triggeredFunction: any;
+      userId?: string;
+      tags?: object;
+      defaultNotificationImageURL?: string;
+    }
+  ) {
+    this.pushwooshInstance.push([
+      'init',
+      {
+        logLevel: 'info',
```
Comment (p-fernandez) [inline providers/pushwoosh/src/lib/pushwoosh.provider.ts:29]: Where does this logLevel value come from?

File: providers/pushwoosh/src/lib/pushwoosh.provider.ts [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+        applicationCode: this.config.applicationCode,
+        defaultNotificationTitle: this.config.defaultNotificationTitle,
+        defaultNotificationImage: this.config.defaultNotificationImageURL,
+        autoSubscribe: this.config.autoSubscribe,
+        subscribeWidget: {
+          enable: true,
+        },
```
Comment (p-fernandez) [inline providers/pushwoosh/src/lib/pushwoosh.provider.ts:36]: What's this property for?

**Review by p-fernandez — CHANGES_REQUESTED** [review]

**Comment by p-fernandez** [comment]
@CursedRock17 sorry for this taking so long to help you to move on. We have changed the branch reference for PR contributions to `next`. This created some conflicts. Could you please update your PR based in the latest version of `next` branch? Apologies for the inconveniences. 🙏🏻

**Comment by CursedRock17** [comment]
@p-fernandez I thought I had addressed all the changes requested, additionally, I updated the branch to slot with the new next branch. Currently though I cannot complete the merge due to errors I can't see.

File: libs/shared/src/consts/providers/provider.enum.ts (hunk 5 above)
Comment (p-fernandez) [reply libs/shared/src/consts/providers/provider.enum.ts:66]: This was a bad suggestion. I am reverting it back. 😓

**Comment by p-fernandez** [comment]
> @p-fernandez I thought I had addressed all the changes requested, additionally, I updated the branch to slot with the new next branch. Currently though I cannot complete the merge due to errors I can't see.

I fixes the merge errors and I rebased your branch into the latest version of our `next` branch so it is up to date.
Now I will review this again to see if we miss anything. Also give it a look to see I didn't remove anything needed from your PR.

File: libs/dal/src/repositories/integration/integration.entity.ts [hunk 10]
```diff
@@ -19,6 +19,13 @@ export interface ICredentials {
   clientId?: string;
   projectName?: string;
   serviceAccount?: string;
+  applicationCode?: string;
+  defaultNotificationTitle?: string;
+  triggeredFunction?: any;
```
Comment (p-fernandez) [inline libs/dal/src/repositories/integration/integration.entity.ts:24]: This is not part of the credentials so it shouldn't be here.

File: libs/shared/src/consts/providers/credentials/provider-credentials.ts [hunk 11]
```diff
@@ -358,6 +358,46 @@ export const expoConfig: IConfigCredentials[] = [
...
+    key: CredentialsKeyEnum.UserId,
+    displayName: 'User ID',
+    type: 'string',
+    required: true,
+  },
+  {
+    key: CredentialsKeyEnum.Tags,
+    displayName: 'Tags',
+    type: 'object',
+    required: true,
```
Comment (p-fernandez) [inline libs/shared/src/consts/providers/credentials/provider-credentials.ts:390]: This is not required based on docs.
https://github.com/Pushwoosh/web-push-notifications#usage

File: libs/shared/src/consts/providers/credentials/provider-credentials.ts [hunk 12]
```diff
@@ -358,6 +358,46 @@ export const expoConfig: IConfigCredentials[] = [
...
+    displayName: 'Auto Subscribe',
+    type: 'boolean',
+    required: true,
+  },
+  {
+    key: CredentialsKeyEnum.UserId,
+    displayName: 'User ID',
+    type: 'string',
+    required: true,
+  },
```
Comment (p-fernandez) [inline libs/shared/src/consts/providers/credentials/provider-credentials.ts:385]: This is not required based on docs.
https://github.com/Pushwoosh/web-push-notifications#usage

File: libs/shared/src/consts/providers/credentials/provider-credentials.ts [hunk 13]
```diff
@@ -358,6 +358,46 @@ export const expoConfig: IConfigCredentials[] = [
...
+    displayName: 'Application Code',
+    type: 'string',
+    required: true,
+  },
+  {
+    key: CredentialsKeyEnum.NotificationTitle,
+    displayName: 'Notification Title',
+    type: 'string',
+    required: true,
+  },
```
Comment (p-fernandez) [inline libs/shared/src/consts/providers/credentials/provider-credentials.ts:373]: This is neither required. Pushwoosh has some defaults that can be overriden.
https://github.com/Pushwoosh/web-push-notifications/blob/401489da1d845c9795383b843f8d3975eab73f13/src/constants.ts#L15

File: libs/shared/src/consts/providers/credentials/provider-credentials.ts [hunk 14]
```diff
@@ -358,6 +358,46 @@ export const expoConfig: IConfigCredentials[] = [
...
+    displayName: 'Tags',
+    type: 'object',
+    required: true,
+  },
+  {
+    key: CredentialsKeyEnum.NotificationImageURL,
+    displayName: 'Notification Image URL',
+    type: 'string',
+    required: true,
+  },
```
Comment (p-fernandez) [inline libs/shared/src/consts/providers/credentials/provider-credentials.ts:397]: This is neither required. Pushwoosh has some defaults that can be overriden.
https://github.com/Pushwoosh/web-push-notifications/blob/401489da1d845c9795383b843f8d3975eab73f13/src/constants.ts#L15

File: libs/shared/src/consts/providers/credentials/provider-credentials.ts [hunk 15]
```diff
@@ -358,6 +358,46 @@ export const expoConfig: IConfigCredentials[] = [
...
+    displayName: 'Notification Title',
+    type: 'string',
+    required: true,
+  },
+  {
+    key: CredentialsKeyEnum.Subscribe,
+    displayName: 'Auto Subscribe',
+    type: 'boolean',
+    required: true,
+  },
```
Comment (p-fernandez) [inline libs/shared/src/consts/providers/credentials/provider-credentials.ts:379]: Seems neither required based on the interface.
https://github.com/Pushwoosh/web-push-notifications/blob/92fcfb43bc6cab273a976f7b41ce5b7a21d01664/types/sdk/pushwoosh.d.ts#L135

File: providers/pushwoosh/src/lib/pushwoosh.provider.spec.ts (hunk 7 above)
Comment (p-fernandez) [inline providers/pushwoosh/src/lib/pushwoosh.provider.spec.ts:3]: Tests need to be added, please. 🙏🏻

File: providers/pushwoosh/src/lib/pushwoosh.provider.ts [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+  constructor(
+    private config: {
+      applicationCode: string;
+      defaultNotificationTitle: string;
+      autoSubscribe: boolean;
+      triggeredFunction: triggeredFunctionType;
+      userId?: string;
+      tags?: object;
+      defaultNotificationImageURL?: string;
+    }
```
Comment (p-fernandez) [inline providers/pushwoosh/src/lib/pushwoosh.provider.ts:26]:
```suggestion
    private config: {
      applicationCode: string;
      defaultNotificationTitle?: string;
      autoSubscribe?: boolean;
      triggeredFunction?: triggeredFunctionType;
      userId?: string;
      tags?: object;
      defaultNotificationImageURL?: string;
    }
```
Based in what I understood from the documents and the API of Pushwoosh.

**Review by p-fernandez — CHANGES_REQUESTED** [review]
I had a look to Pushwoosh documentation and API. We need to address some changes, please.

File: providers/pushwoosh/src/lib/pushwoosh.provider.ts [hunk 17]
```diff
@@ -0,0 +1,57 @@
+import {
+  ChannelTypeEnum,
+  ISendMessageSuccessResponse,
+  IPushOptions,
+  IPushProvider,
+} from '@novu/stateless';
+
+import { Pushwoosh } from 'web-push-notifications';
```
Comment (scopsy) [inline providers/pushwoosh/src/lib/pushwoosh.provider.ts:8]: A small question @CursedRock17 is there a particular reason why are we using the client library here?
I've came across a node client https://www.npmjs.com/package/pushwoosh-client should we perhaps switch to it.

File: providers/pushwoosh/src/lib/pushwoosh.provider.ts [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+  }
+
+  async sendMessage(
+    options: IPushOptions
+  ): Promise<ISendMessageSuccessResponse> {
+    await this.pushwooshInstance.push(function (api) {
+      this.pushwooshInstance.addEventHandler(options.title, function (payload) {
+        this.config.triggeredFunction(payload);
+      });
+    });
```
Comment (scopsy) [inline providers/pushwoosh/src/lib/pushwoosh.provider.ts:50]: The combination here with a callback will not really be performed in the anticipated order from what i see. All the inner contents are callback based which will execute it later in the flow.

**Review by scopsy — CHANGES_REQUESTED** [review]
Added a few notes here regarding the SDK. My main concern is that we are using a web SDK instead of a backend one. I'm afraid it will create issues with out multi tenancy nature. @CursedRock17 are you familiar with the pushwoosh node SDK? Any particular reason of not using it?

**Comment by scopsy** [comment]
Closing this for now due to inactivity, feel free to reopen this once you have a moment. Thank you a lot for the contribution!

**Pull request closed without merging**